core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -925,47 +933,156 @@ public boolean isHashValid(String hash) {
+            byte[] hash = new byte[0];
+            try {
+                secretKeyFactory = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+                hash = secretKeyFactory.generateSecret(spec).getEncoded();
+            } catch (NoSuchAlgorithmException e) {
+                LOGGER.log(WARNING, "No such algorithm found for encode",  e);
+            } catch (InvalidKeySpecException e) {
+                LOGGER.log(WARNING, "Invalid key spec exception",  e);
+            }
+            return hash;
+        }
+
+        private static String toHex(byte[] secretKey)  {
+
+            BigInteger bi = new BigInteger(1, secretKey);
+            String saltHashValue = bi.toString(RADIX_LENGTH);
+
+            int paddingLength = (secretKey.length * HEX_TWO) - saltHashValue.length();
+            if (paddingLength > ZERO) {
+                return String.format("%0" + paddingLength + "d", ZERO) + saltHashValue;
+            } else {
+                return saltHashValue;
+            }
+        }
+
+        @SuppressFBWarnings(value = {"DMI_RANDOM_USED_ONLY_ONCE", "PREDICTABLE_RANDOM"}, justification = "https://github.com/spotbugs/spotbugs/issues/1539 and doesn't need to be secure, we're just not hardcoding a 'wrong' password")
+        private static byte[] generateSalt() {
+            SecureRandom random = new SecureRandom();
+            byte[] salt = new byte[16];
+            random.nextBytes(salt);
+            return salt;
+        }
+
+        @Override
+        public boolean isHashValid(String hash)    {
+             return true;
+        }
+
+        private static boolean validatePassword(String originalPassword, String storedPassword) {
+            String[] parts = storedPassword.split(STRING_SEPARATION);
+            int iterations = Integer.parseInt(parts[0]);
+
+            byte[] salt = fromHex(parts[1]);
+            byte[] hash = fromHex(parts[2]);
+
+            PBEKeySpec spec = new PBEKeySpec(originalPassword.toCharArray(),
+                    salt, iterations, hash.length * HASH_LENGTH);
+
+            byte[] generatedHashValue = generateSecretKey(spec);
+
+            int diff = hash.length ^ generatedHashValue.length;
+            for (int i = ZERO; i < hash.length && i < generatedHashValue.length; i++)
+            {
+                diff |= hash[i] ^ generatedHashValue[i];
+            }
+            return diff == 0;
+        }
+
+        private static byte[] fromHex(String storedPassword)  {
+            byte[] bytes = new byte[storedPassword.length() / HEX_TWO];
+            for (int i = ZERO; i < bytes.length; i++)
+            {
+                bytes[i] = (byte) Integer.parseInt(storedPassword.substring(HEX_TWO * i, HEX_TWO * i + HEX_TWO), RADIX_LENGTH);
+            }
+            return bytes;
+        }
+
+    }
+
+    /* package */ static final PasswordHashEncoder PASSWORD_HASH_ENCODER =  Util.isFipsMode() ? new PBKDF2PasswordEncoder() : new JBCryptEncoder();
```

Comment:
will be useful if we log the encoder being picked up at runtime, will be helpful during triage.